core/system/src/export_table.c

```diff
@@ -83,16 +95,446 @@ static void thread_destroy(void * c)
+                semaphore_post(&fn_group->semaphore);
+                ++num_posted;
+            }
+        }
+
+        slot = fn_group_queue->pool->management_array[slot].queued.next;
+    }
+
+    return num_posted;
+}
+
+static int callee_box_id(const TRPCGateway * gateway)
+{
+    return (uint32_t *)gateway->box_ptr - __uvisor_config.cfgtbl_ptr_start;
+}
+
+static int put_it_back(uvisor_pool_queue_t * queue, uvisor_pool_slot_t slot)
+{
+    int status;
+    status = uvisor_pool_queue_try_enqueue(queue, slot);
+    if (status) {
+        /* We could dequeue an RPC message, but couldn't put it back. */
+        /* It is bad to take down the entire system. It is also bad
+         * to lose messages due to not being able to put them back in
+         * the queue. However, if we could dequeue the slot
+         * we should have no trouble enqueuing the slot here. */
+        assert(false);
+    }
+
+    /* Note that we don't have to modify the message in the queue, since it'll
+     * still be valid. Nobody else will have run at the same time that could
+     * have messed it up. */
+
+     return status;
+}
+
+/* Return true iff gateway is valid. */
+static int is_valid_rpc_gateway(const TRPCGateway * const gateway)
+{
+    /* Gateway needs to be entirely in flash. */
+    uint32_t gateway_start = (uint32_t) gateway;
+    uint32_t gateway_end = gateway_start + sizeof(*gateway);
+    if (!(vmpu_public_flash_addr(gateway_start) && vmpu_public_flash_addr(gateway_end))) {
+        return 0;
+    }
+
+    /* The callee box ID associated with the gateway must be valid. Gateways to
+     * box 0 are not possible. */
+    int box_id = callee_box_id(gateway);
+    if (box_id <= 0 || box_id >= g_vmpu_box_count) {
+        return 0;
+    }
+
+    /* Gateway needs to have good magic. */
+    if (!(gateway->magic == UVISOR_RPC_GATEWAY_MAGIC_ASYNC || gateway->magic == UVISOR_RPC_GATEWAY_MAGIC_SYNC)) {
+        return 0;
+    }
+
+    /* Gateway needs good ldr_pc instruction. */
+    if (gateway->ldr_pc != LDR_PC_PC_IMM_OPCODE(__UVISOR_OFFSETOF(TRPCGateway, ldr_pc),
+                                                __UVISOR_OFFSETOF(TRPCGateway, caller))) {
+        return 0;
+    }
+
+    /* Gateway needs to point to functions in flash (caller and target) */
+    if (!(vmpu_public_flash_addr(gateway->caller) && vmpu_public_flash_addr(gateway->target))) {
+        return 0;
+    }
+
+    /* All checks passed. */
+    return 1;
+}
+
+/* Return true if and only if the queue is entirely within the box specified by
+ * the provided box_id. */
+static int is_valid_queue(uvisor_pool_queue_t * queue, int box_id)
+{
+    uint32_t bss_start = g_context_current_states[box_id].bss;
+    uint32_t bss_end = bss_start + g_context_current_states[box_id].bss_size;
+
+    uint32_t queue_start = (uint32_t) queue;
+    uint32_t queue_end = queue_start + sizeof(*queue);
+    int queue_is_valid = queue_start >= bss_start && queue_end <= bss_end;
```

Comment:
I don't think the magic is really necessary. We can use the queue directly from the box config table instead of from the box context pointer. The box cfg magic is already checked. We can avoid these sorts of errors from that box cfg magic check. We also don't have to trust the box context pointer to the queue, although we still need to validate the queue itself.